### PR TITLE
Add some inference annotations

### PR DESCRIPTION
**Pull request by nickdrozd**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

### CODE REVIEW

File: astroid/decorators.py [hunk 1]
```diff
@@ -89,15 +89,17 @@ def __get__(self, inst, objtype=None):
         return val
 
 
-def path_wrapper(func):
+def path_wrapper(func: Callable) -> Callable:
```
Comment (DanielNoord) [inline astroid/decorators.py:92]: This should probably be a `Typevar` or even a `ParamSpec`.

File: astroid/helpers.py [hunk 2]
```diff
@@ -219,7 +219,7 @@ def is_supertype(type1, type2) -> bool:
     return _type_check(type1, type2)
 
 
-def class_instance_as_index(node):
+def class_instance_as_index(node) -> nodes.NodeNG | None:
```
Comment (DanielNoord) [inline astroid/helpers.py:222]:
```suggestion
def class_instance_as_index(node) -> nodes.Const | None:
```

?

File: astroid/inference.py [hunk 3]
```diff
@@ -80,14 +80,16 @@ def infer_end(
...
+nodes.ClassDef._infer = infer_end
 nodes.Lambda._infer = infer_end  # type: ignore[assignment]
 nodes.Const._infer = infer_end  # type: ignore[assignment]
 nodes.Slice._infer = infer_end  # type: ignore[assignment]
 
 
-def _infer_sequence_helper(node, context: InferenceContext | None = None):
+def _infer_sequence_helper(
+    node: _BaseContainerT, context: InferenceContext | None = None
+) -> list[nodes.NodeNG]:
```
Comment (DanielNoord) [inline astroid/inference.py:92]: Is `node.elts` correctly typed?

File: astroid/inference.py [hunk 4]
```diff
@@ -193,7 +195,7 @@ def _infer_map(
             if any(not elem for elem in (key, safe_value)):
                 raise InferenceError(node=node, context=context)
             # safe_value is SuccessfulInferenceResult as bool(Uninferable) == False
-            values = _update_with_replacement(values, {key: safe_value})  # type: ignore[dict-item]
+            values = _update_with_replacement(values, {key: safe_value})
```
Comment (DanielNoord) [inline astroid/inference.py:198]: Have you tested this with ``mypy --strict``? I added these ignores based on strict mode so they might not be unused in that case.

File: astroid/inference.py [hunk 5]
```diff
@@ -772,7 +776,7 @@ def _get_binop_flow(
     right_type: InferenceResult | None,
     context: InferenceContext,
     reverse_context: InferenceContext,
-):
+) -> list:
```
Comment (DanielNoord) [inline astroid/inference.py:779]: I have PRs for all binary inference code paths after https://github.com/PyCQA/astroid/pull/1787. Let's fix those after that PR gets merged.

File: astroid/inference.py [hunk 6]
```diff
@@ -1139,10 +1143,12 @@ def infer_empty_node(
 nodes.EmptyNode._infer = infer_empty_node  # type: ignore[assignment]
 
 
-def _populate_context_lookup(call, context):
+def _populate_context_lookup(
+    call: nodes.Call, context: InferenceContext | None
+) -> dict[nodes.NodeNG, InferenceContext]:
```
Comment (DanielNoord) [inline astroid/inference.py:1148]: This feels like it should be `InferenceResult` or something.

File: astroid/interpreter/dunder_lookup.py [hunk 7]
```diff
@@ -18,7 +18,7 @@
 from astroid.exceptions import AttributeInferenceError
 
 
-def _lookup_in_mro(node, name):
+def _lookup_in_mro(node, name) -> list:
```
Comment (DanielNoord) [inline astroid/interpreter/dunder_lookup.py:21]: Can we already try narrowing this down?

File: astroid/nodes/_base_nodes.py [hunk 8]
```diff
@@ -150,7 +150,7 @@ def do_import_module(self, modname: str | None = None) -> nodes.Module:
             use_cache=use_cache,
         )
 
-    def real_name(self, asname):
+    def real_name(self, asname) -> str:
```
Comment (DanielNoord) [inline astroid/nodes/_base_nodes.py:153]:
```suggestion
    def real_name(self, asname: str) -> str:
```
?

File: astroid/protocols.py [hunk 9]
```diff
@@ -349,7 +349,7 @@ def assend_assigned_stmts(
 nodes.AssignAttr.assigned_stmts = assend_assigned_stmts
 
 
-def _arguments_infer_argname(self, name, context):
+def _arguments_infer_argname(self, name: sttr, context: InferenceContext) -> Iterator:
```
Comment (DanielNoord) [inline astroid/protocols.py:352]: There are return statements here so should we use `Generator`?

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Initial pass!

File: astroid/inference.py (hunk 3 above)
Comment (nickdrozd) [reply astroid/inference.py:92]: `BaseContainer` has `self.elts: list[NodeNG] = []`

File: astroid/inference.py (hunk 4 above)
Comment (nickdrozd) [reply astroid/inference.py:198]: Yes, with version `mypy 0.991 (compiled: yes)` and `Python 3.8.10`. It warned that only some of the ignores are unused. I guess some have the nodes have the function types declared and some don't?

File: astroid/interpreter/dunder_lookup.py (hunk 7 above)
Comment (nickdrozd) [reply astroid/interpreter/dunder_lookup.py:21]: I'm not sure what it's supposed to be.

File: astroid/decorators.py (hunk 1 above)
Comment (nickdrozd) [reply astroid/decorators.py:92]: Could you explain more? I haven't used those.

File: astroid/inference.py (hunk 6 above)
Comment (nickdrozd) [reply astroid/inference.py:1148]: Which part?

File: astroid/decorators.py (hunk 1 above)
Comment (DanielNoord) [reply astroid/decorators.py:92]: I think the PEP introducing `ParamSpec` is quite clear:
https://peps.python.org/pep-0612/

It explains its uses and those can be applied here.

File: astroid/inference.py (hunk 3 above)
Comment (DanielNoord) [reply astroid/inference.py:92]: Hmm that sounds weird. I need to investigate this but I'm fairly certain that it could be `Instance` as well.

File: astroid/inference.py (hunk 6 above)
Comment (DanielNoord) [reply astroid/inference.py:1148]: The `nodes.NodeNG`

File: astroid/decorators.py [hunk 10]
```diff
@@ -89,15 +89,17 @@ def __get__(self, inst, objtype=None):
...
 
     Used to stop inference if the node has already been looked
     at for a given `InferenceContext` to prevent infinite recursion
     """
 
     @functools.wraps(func)
-    def wrapped(node, context: InferenceContext | None = None, _func=func, **kwargs):
+    def wrapped(
+        node, context: InferenceContext | None = None, _func=func, **kwargs
+    ) -> Iterator:
```
Comment (DanielNoord) [inline astroid/decorators.py:102]: Let's use `Generator`, since this function has `return` statements.

File: astroid/helpers.py [hunk 11]
```diff
@@ -219,7 +219,7 @@ def is_supertype(type1, type2) -> bool:
     return _type_check(type1, type2)
 
 
-def class_instance_as_index(node):
+def class_instance_as_index(node) -> nodes.Const | None:
```
Comment (DanielNoord) [inline astroid/helpers.py:222]:
```suggestion
def class_instance_as_index(node: bases.Instance) -> nodes.Const | None:
```

I did a quick check and this should work 😄

File: astroid/inference.py (hunk 3 above)
Comment (DanielNoord) [reply astroid/inference.py:92]: I checked and this is likely incorrect. In the construction of `Lists` in `DictModel.attr_values` we use the `keys` of a `Dict.items`. Those are `SuccessfulInferenceResults`.
So `elts` is indeed incorrectly typed.

File: astroid/inference.py [hunk 12]
```diff
@@ -216,7 +218,7 @@ def _higher_function_scope(node: nodes.NodeNG) -> nodes.FunctionDef | None:
     while current.parent and not isinstance(current.parent, nodes.FunctionDef):
         current = current.parent
     if current and current.parent:
-        return current.parent  # type: ignore[return-value]
+        return current.parent
```
Comment (DanielNoord) [inline astroid/inference.py:221]: This causes an additional message.

File: astroid/inference.py [hunk 13]
```diff
@@ -743,7 +745,9 @@ def _bin_op(
     )
 
 
-def _get_binop_contexts(context, left, right):
+def _get_binop_contexts(
```
Comment (DanielNoord) [inline astroid/inference.py:748]: Let's do this in my `BinOp` PRs. I'll prepare those somewhere this week.

File: astroid/inference.py [hunk 14]
```diff
@@ -759,7 +763,7 @@ def _get_binop_contexts(context, left, right):
         yield new_context
 
 
-def _same_type(type1, type2):
+def _same_type(type1, type2) -> bool:
```
Comment (DanielNoord) [inline astroid/inference.py:766]: This is interesting as adding annotations to `type1` will cause a type error... Hmm, we should probably investigate this.

File: astroid/protocols.py [hunk 15]
```diff
@@ -349,7 +349,9 @@ def assend_assigned_stmts(
 nodes.AssignAttr.assigned_stmts = assend_assigned_stmts
 
 
-def _arguments_infer_argname(self, name, context):
+def _arguments_infer_argname(
+    self, name: str, context: InferenceContext
```
Comment (DanielNoord) [inline astroid/protocols.py:353]: `name` can be `None`, see code down below.

File: pyproject.toml [hunk 16]
```diff
@@ -78,6 +78,8 @@ no_implicit_optional = true
 scripts_are_modules = true
 show_error_codes = true
 warn_redundant_casts = true
+allow_untyped_decorators = true
```
Comment (DanielNoord) [inline pyproject.toml:81]: Let's do this in another PR, if we really want to do this. But let's keep the discussion in that PR.

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Did another pass! Thanks for taking the time to do this, I know how difficult it is to get these right and PRs like this are really helpful to at least get the discussion and project going 😄

File: astroid/inference.py (hunk 14 above)
Comment (nickdrozd) [reply astroid/inference.py:766]: This function is used twice. In both cases, the inputs can be `InferenceResult | None`, but this function assumes that it will be `InferenceResult`.

Comment (nickdrozd) [reply astroid/inference.py:766]: `Proxy` is involved too. Messy!

File: astroid/decorators.py (hunk 1 above)
Comment (nickdrozd) [reply astroid/decorators.py:92]: That sounds out of scope for this PR. I can undo this change if you want.

File: astroid/inference.py (hunk 3 above)
Comment (nickdrozd) [reply astroid/inference.py:92]: Should this change be undone?

File: astroid/decorators.py (hunk 1 above)
Comment (DanielNoord) [reply astroid/decorators.py:92]: Yeah, let's revert then

File: astroid/inference.py (hunk 3 above)
Comment (DanielNoord) [reply astroid/inference.py:92]: Can't you change it to `SuccessfulInferenceResult`?

File: astroid/inference.py (hunk 14 above)
Comment (DanielNoord) [reply astroid/inference.py:764]: 😅

File: astroid/inference.py [hunk 17]
```diff
@@ -1139,10 +1141,12 @@ def infer_empty_node(
 nodes.EmptyNode._infer = infer_empty_node  # type: ignore[assignment]
 
 
-def _populate_context_lookup(call, context):
+def _populate_context_lookup(
+    call: nodes.Call, context: InferenceContext | None
+) -> dict[InferenceResult, InferenceContext]:
```
Comment (DanielNoord) [inline astroid/inference.py:1146]: Have you actually checked that this is `InferenceResult` and not `SuccesfulInferenceResult`?

**Review by DanielNoord — CHANGES_REQUESTED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
Looks pretty good, there are a lot of fixes. I would be in favor of merging. We definitely need to activate mypy to be able to tell if something regressed without having to check locally by comparing the output before and after (I didn't do that so the approve is based on trust and the fact that actually verifying would take a lot of time.)

**Review by DanielNoord — APPROVED** [review]
I did verify this last time and all comments have been resolved! Thanks!
